Added X/Y inputs to design panel

barklund opened the pull request:
## Summary

This adds a new row of inputs to the size&position design panel.

**NB:** This comes with some very important and pertinent context:

* This is added for accessibility reasons - it makes it easier to place elements when using keyboard-only.
* This is a very temporary solution. There's a new design coming and these inputs haven't been designed in the old (existing design). We're just adding them here for stable, but what we really want is the new inputs in new design which'll happen post stable.
* Min-max limits are a bit up in the air (currently e.g. `[-width,PAGE_WIDTH]` for x) and there's no automatic removal of element if moved off-stage in this way. I didn't feel that (removing if off stage through design panel) was a good result if you accidentally typed a `0` too much.
* I will not be adding tests, as this is temporary.

Temporary design looks like this:
![Screen Shot 2020-08-20 at 23 06 52](https://user-images.githubusercontent.com/637548/90848047-e736f100-e339-11ea-8aff-8e7a20388f18.png)

## Relevant Technical Choices

* I added some alignment of the single-character suffix to have the single letter center-aligned - it looked better.
* I also added some dummy spacing between the two elements to align with the width/height ones.

## To-do

N/A

## User-facing changes

* There's a new row in the size&position panel for non-bg elements allowing the creator to set the position of elements.

## Testing Instructions

* If there's a new row in the size&position panel for non-bg elements, it's the new work.

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #3636

Commit:
Added new position inputs in existing design to size&position panel

Commit:
Added spacer to better align inputs

Commit:
Made nicer alignment of single-char suffixes

Commit:
Allowed negative inputs

Commit:
Fixed panel tests so only relevant props are matched

miina commented on assets/src/edit-story/components/panels/sizePosition.js:
```diff
@@ -62,12 +63,26 @@ const MIN_MAX = {
   },
+  X: {
+    MIN: 0,
+    MAX: PAGE_WIDTH,
```
Maybe these should be `1` and `PAGE_WIDTH - 1` or sth like this instead to ensure it stays on the Page? Otherwise, the element won't be visible.

merapi replied:
`X.MIN = 0` is good, I can place the element on canvas like that (it snaps).
`X.MAX = PAGE_WIDTH` is debatable, it will not be visible on the page (but not removed).

![image](https://user-images.githubusercontent.com/21036224/90871243-087bf980-e39b-11ea-9f15-501f11766072.png)

miina replied:
@merapi Note that in usage it will actually be `MIN_MAX.X.MIN - width`, which means `0 - width` => the element will be out of the page.

merapi replied:
Yeah, for the start we need to decide what to do with that one pixel (is it out of page?).

barklund replied:
Fixed in d93e1823f.

miina commented on assets/src/edit-story/components/panels/sizePosition.js:
```diff
@@ -62,12 +63,26 @@ const MIN_MAX = {
+    MAX: PAGE_WIDTH,
+  },
+  Y: {
+    MIN: 0,
```
Note that `0` for `y` is actually the safezone limit. So the very top of the Page is `0 - [SAFE_ZONE_AREA]`

barklund replied:
Fixed in d93e1823f.

miina commented on assets/src/edit-story/components/panels/sizePosition.js:
```diff
@@ -238,6 +265,28 @@ function SizePositionPanel({
+          value={x}
+          min={MIN_MAX.X.MIN - width}
+          max={MIN_MAX.X.MAX}
+          onChange={(value) => pushUpdate({ x: value })}
```
Would be great if the element would actually move on the Page without having to blur first, especially when using keyboard only. That seems to be a general issue though, maybe we could have a debounced callback in the background saving the pending updates. For a future PR though.

barklund replied:
Yeah, let's delay that until the coming complete overhaul of design panels, that'll happen post-stable anyway.

merapi commented on assets/src/edit-story/components/panels/sizePosition.js:
```diff
@@ -238,6 +265,28 @@ function SizePositionPanel({
+        <BoxedNumeric
+          suffix={_x('Y', 'Position on Y axis', 'web-stories')}
+          value={y}
+          min={MIN_MAX.Y.MIN - height}
```
This is connected to @miina comment about `0-safezone`, I think we can leave `Y.MIN=0` but here we need to account for the safezone.

Editor:
![image](https://user-images.githubusercontent.com/21036224/90872450-f0a57500-e39c-11ea-9094-d186d63d2730.png)

Output:
![image](https://user-images.githubusercontent.com/21036224/90872512-0b77e980-e39d-11ea-9e2c-630b9704e71e.png)

merapi replied:
And we have this? (or this+1 so the element is visible)
![image](https://user-images.githubusercontent.com/21036224/90872965-acff3b00-e39d-11ea-954b-d27860c2709b.png)

barklund replied:
Fixed in d93e1823f.

Commit:
Fixed min values for position inputs

Commit:
Fixed max value for Y position

miina commented on assets/src/edit-story/components/panels/sizePosition.js:
```diff
@@ -238,6 +266,28 @@ function SizePositionPanel({
+        <BoxedNumeric
+          suffix={_x('X', 'Position on X axis', 'web-stories')}
+          value={x}
+          min={MIN_MAX.X.MIN - width}
```
Hmm, just realized that for the limits to actually make sense, it should also consider rotation when subtracting width/height...

miina replied:
Otherwise this would be allowed for example:
<img width="894" alt="Screenshot 2020-08-24 at 12 00 10" src="https://user-images.githubusercontent.com/3294597/91025022-10d17000-e5f9-11ea-924a-41a3011d9bfd.png">

Thoughts?

merapi replied:
True. While X&Y coords are not so easy to operate when rotation is involved - you need to do some math before you put something in the input, but when using a keyboard, the min/max limit would be nice.

BTW: Checked how Figma does this (different from us, they update the coords):
Box at 0x0
![image](https://user-images.githubusercontent.com/21036224/91169602-edc9bd80-e6d7-11ea-8bb5-4e64b693b1c5.png)

Coords are updated when rotation is applied
![image](https://user-images.githubusercontent.com/21036224/91169628-f8845280-e6d7-11ea-9aaf-7bb16ee678d0.png)

Move to 0x0 (not as some would like and not as we do it)
![image](https://user-images.githubusercontent.com/21036224/91169673-06d26e80-e6d8-11ea-9e58-dbc00966aad8.png)

barklund replied:
Yeah, there was this little tiny voice and the back of my head screaming this over and over but I managed to suppress it all this time. I've addressed it now in fe8118780.

barklund replied:
> BTW: Checked how Figma does this (different from us, they update the coords)

Yeah, that's left as an exercise for future improvement. It makes a lot of sense, that the input for x is always the left-most edge of an element, but it's not easy.

In this rewrite I am actually getting the information necessary, so we could simply translate it back and forth in the input only (and still use a single anchor corner internally), but it might come with other errors.

Commit:
Merge branch 'main' into feature/3636-add-position-panel

Commit:
Use rotated size and position in calcs

Note that this file is growing out of hand quickly, however that will be dealt with in the coming refactor to new design in #4104 - no need to worry yet!

Commit:
Merge branch 'main' into feature/3636-add-position-panel
